fdbserver/BlobManager.actor.cpp

```diff
@@ -1911,19 +2090,43 @@ ACTOR Future<Void> granuleMergeChecker(Reference<BlobManagerData> bmData) {
 		auto allRanges = bmData->mergeCandidates.ranges();
 		std::vector<std::tuple<UID, KeyRange, Version>> currentCandidates;
 
+		auto& mergeBoundaries = bmData->mergeBoundaries;
 		for (auto& it : allRanges) {
-			if (!it->cvalue().canMergeNow() || currentCandidates.size() == maxRangeSize) {
+			// Conditions:
+			//   1. Next range is not eligible.
+			//   2. Hit the maximum in a merge evaluation window.
+			//   3. Hit a hard merge boundary meaning we should not merge across them.
+			if (!it->cvalue().mergeEligible() || currentCandidates.size() == maxRangeSize ||
+				bmData->mergeHardBoundaries.count(it.begin())) {
 				if (currentCandidates.size() >= 2) {
 					mergeChecks.push_back(attemptMerges(bmData, currentCandidates));
 				}
 				currentCandidates.clear();
 			}
 
-			if (it->cvalue().canMergeNow()) {
+			// Soft boundaries.
+			// We scan all ranges including non-eligible ranges which trigger non-consecutive merge flushes.
+			if (!currentCandidates.empty()) {
+				KeyRangeRef lastRange = std::get<1>(currentCandidates.back());
+				KeyRangeRef curRange = it->range();
+				ASSERT(lastRange.end == curRange.begin);
+				// Conditions:
+				//   1. Start a new soft merge range.
+				//   2. End a soft merge range.
+				if ((!mergeBoundaries.count(curRange.begin) && mergeBoundaries.count(curRange.end)) ||
```

Comment:
We could also do a similar optimization here to avoid starting a candidate set on a merge boundary, of having a parallel iterator over mergeBoundaries, instead of doing a count for each one.

Reply:
Hmmm. I don't think there's anything wrong with starting a candidate set on a merge boundary. Say we have A -> A | B -> A | B | C, where B -> B | C. There's nothing wrong with A | B | C -> A | B again. Maybe I'm missing what you're saying.